File: src/Application Specific/Classes/Structures/SubMenuNode.vb
Public Class SubMenuNode : Inherits LinkLabel

#Region "Constructors"

    Public Sub New(ByVal nameIn As String)
        LevelNumber = 1
        MyName = nameIn
    End Sub

    Public Sub New(ByVal nameIn As String, ByVal LevelNumberIn As Integer)
        LevelNumber = LevelNumberIn
        MyName = nameIn
    End Sub

    Public Sub New(ByVal nameIn As String, ByVal tableToSearchIn As Entity.Sys.Enums.TableNames)
        LevelNumber = 1
        MyName = nameIn
        TableToSearch = tableToSearchIn
    End Sub

    Public Sub New(ByVal nameIn As String, ByVal tableToSearchIn As Entity.Sys.Enums.TableNames, ByVal LevelNumberIn As Integer)
        LevelNumber = LevelNumberIn
        MyName = nameIn
        TableToSearch = tableToSearchIn
    End Sub

    Public Sub New(ByVal nameIn As String, ByVal formToOpenIn As String)
        LevelNumber = 1
        MyName = nameIn
        FormToOpen = formToOpenIn
    End Sub

    Public Sub New(ByVal nameIn As String, ByVal formToOpenIn As String, ByVal LevelNumberIn As Integer)
        LevelNumber = LevelNumberIn
        MyName = nameIn
        FormToOpen = formToOpenIn
    End Sub

#End Region

#Region "Properties"

    Private _LevelNumber As Integer = 1
    Public Property LevelNumber() As Integer
        Get
            Return _LevelNumber
        End Get
        Set(ByVal Value As Integer)
            _LevelNumber = Value
        End Set
    End Property

    Private _MyName As String = ""
    Public Property MyName() As String
        Get
            Return _MyName
        End Get
        Set(ByVal Value As String)
            _MyName = Value
            Me.Text = MyName

            If LevelNumber = 1 Then
                Me.Location = New Point(5, MainForm.MenuBar.pnlSubMenu.Controls.Count * 20)
                Me.Size = New Size(152, 15)
            Else
                Me.Location = New Point(10 * LevelNumber, MainForm.MenuBar.pnlSubMenu.Controls.Count * 20)
                Me.Size = New Size(152 - (10 * LevelNumber), 15)
            End If

            Me.LinkColor = Color.FromArgb(0, 52, 102)
            Me.ActiveLinkColor = Color.FromArgb(0, 52, 102)
            Me.DisabledLinkColor = Color.FromArgb(0, 52, 102)
            Me.VisitedLinkColor = Color.FromArgb(0, 52, 102)

            AddHandler Me.Click, AddressOf ButtonClicked
        End Set
    End Property

    Private _ChildNodes As Array = Nothing
    Public Property ChildNodes() As Array
        Get
            Return _ChildNodes
        End Get
        Set(ByVal Value As Array)
            _ChildNodes = Value

            For Each childNode As SubMenuNode In ChildNodes
                MainForm.MenuBar.pnlSubMenu.Controls.Add(childNode)
            Next
        End Set
    End Property

    Private _FormToOpen As String = "UsePanels"
    Public Property FormToOpen() As String
        Get
            Return _FormToOpen
        End Get
        Set(ByVal Value As String)
            _FormToOpen = Value
        End Set
    End Property

    Private _TableToSearch As Entity.Sys.Enums.TableNames = Entity.Sys.Enums.TableNames.NO_TABLE
    Public Property TableToSearch() As Entity.Sys.Enums.TableNames
        Get
            Return _TableToSearch
        End Get
        Set(ByVal Value As Entity.Sys.Enums.TableNames)
            _TableToSearch = Value
        End Set
    End Property

#End Region

#Region "Functions"

    Private Sub ButtonClicked(ByVal sender As Object, ByVal e As EventArgs)
        Navigation.Sub_Menu(Me)
    End Sub

#End Region

End Class

File: apollo/apollo.fsm/src/Application Specific/Classes/Structures/SubMenuNode.cs
using System;
using System.Drawing;
using System.Windows.Forms;

namespace FSM
{
    public class SubMenuNode : LinkLabel
    {
        public SubMenuNode(string nameIn)
        {
            LevelNumber = 1;
            MyName = nameIn;
        }

        public SubMenuNode(string nameIn, int LevelNumberIn)
        {
            LevelNumber = LevelNumberIn;
            MyName = nameIn;
        }

        public SubMenuNode(string nameIn, Entity.Sys.Enums.TableNames tableToSearchIn)
        {
            LevelNumber = 1;
            MyName = nameIn;
            TableToSearch = tableToSearchIn;
        }

        public SubMenuNode(string nameIn, Entity.Sys.Enums.TableNames tableToSearchIn, int LevelNumberIn)
        {
            LevelNumber = LevelNumberIn;
            MyName = nameIn;
            TableToSearch = tableToSearchIn;
        }

        public SubMenuNode(string nameIn, string formToOpenIn)
        {
            LevelNumber = 1;
            MyName = nameIn;
            FormToOpen = formToOpenIn;
        }

        public SubMenuNode(string nameIn, string formToOpenIn, int LevelNumberIn)
        {
            LevelNumber = LevelNumberIn;
            MyName = nameIn;
            FormToOpen = formToOpenIn;
        }

        private int _LevelNumber = 1;

        public int LevelNumber
        {
            get
            {
                return _LevelNumber;
            }

            set
            {
                _LevelNumber = value;
            }
        }

        private string _MyName = "";

        public string MyName
        {
            get
            {
                return _MyName;
            }

            set
            {
                _MyName = value;
                Text = MyName;
                if (LevelNumber == 1)
                {
                    Location = new Point(5, App.MainForm.MenuBar.pnlSubMenu.Controls.Count * 20);
                    Size = new Size(152, 15);
                }
                else
                {
                    Location = new Point(10 * LevelNumber, App.MainForm.MenuBar.pnlSubMenu.Controls.Count * 20);
                    Size = new Size(152 - 10 * LevelNumber, 15);
                }

                LinkColor = Color.FromArgb(0, 52, 102);
                ActiveLinkColor = Color.FromArgb(0, 52, 102);
                DisabledLinkColor = Color.FromArgb(0, 52, 102);
                VisitedLinkColor = Color.FromArgb(0, 52, 102);
                Click += ButtonClicked;
            }
        }

        private Array _ChildNodes = null;

        public Array ChildNodes
        {
            get
            {
                return _ChildNodes;
            }

            set
            {
                _ChildNodes = value;
                foreach (SubMenuNode childNode in ChildNodes)
                    App.MainForm.MenuBar.pnlSubMenu.Controls.Add(childNode);
            }
        }

        private string _FormToOpen = "UsePanels";

        public string FormToOpen
        {
            get
            {
                return _FormToOpen;
            }

            set
            {
                _FormToOpen = value;
            }
        }

        private Entity.Sys.Enums.TableNames _TableToSearch = Entity.Sys.Enums.TableNames.NO_TABLE;

        public Entity.Sys.Enums.TableNames TableToSearch
        {
            get
            {
                return _TableToSearch;
            }

            set
            {
                _TableToSearch = value;
            }
        }

        private void ButtonClicked(object sender, EventArgs e)
        {
            Navigation.Sub_Menu(this);
        }
    }
}